apollo-router/src/query_planner/fetch.rs

```diff
@@ -243,6 +255,189 @@ impl Display for QueryHash {
+    let old_query = "query QueryLL__Subgraph1__1($representations:[_Any!]!$Subgraph1_U_field_a:String!){_entities(representations:$representations){...on U{id field(a:$Subgraph1_U_field_a)}}}";
+    let mut ctx_args = HashSet::new();
+    ctx_args.insert("Subgraph1_U_field_a".to_string());
+    let contextual_args = Some((ctx_args, 2));
+
+    let expected = "query QueryLL__Subgraph1__1($representations: [_Any!]!, $Subgraph1_U_field_a_0: String!, $Subgraph1_U_field_a_1: String!) { _0: _entities(representations: $representations) { ... on U { id field(a: $Subgraph1_U_field_a_0) } } _1: _entities(representations: $representations) { ... on U { id field(a: $Subgraph1_U_field_a_1) } } }";
+
+    assert_eq!(
+        expected,
+        query_batching_for_contextual_args(old_query, &contextual_args)
+            .unwrap()
+            .unwrap()
+    );
+}
+
+// TODO: There is probably a function somewhere else that already does this
+fn data_at_path<'v>(data: &'v Value, path: &Path) -> Option<&'v Value> {
+    let v = match &path.0[0] {
+        PathElement::Fragment(s) => {
+            // get the value at data.get("__typename") and compare it with s. If the values are equal, return data, otherwise None
+            let mut z: Option<&Value> = None;
+            let wrapped_typename = data.get("__typename");
+            if let Some(t) = wrapped_typename {
+                if t.as_str() == Some(s.as_str()) {
+                    z = Some(data);
+                }
+            }
+            z
+        }
+        PathElement::Key(v, _) => data.get(v),
+        PathElement::Index(idx) => Some(&data[idx]),
+        PathElement::Flatten(_) => None,
+    };
+
+    if path.len() > 1 {
+        if let Some(val) = v {
+            let remaining_path = path.iter().skip(1).cloned().collect();
+            return data_at_path(val, &remaining_path);
+        }
+    }
+    v
+}
+
+fn merge_context_path(current_dir: &Path, context_path: &Path) -> Path {
+    let mut i = 0;
+    let mut j = current_dir.len();
+    // iterate over the context_path(i), every time we encounter a '..', we want
+    // to go up one level in the current_dir(j)
+    while i < context_path.len() {
+        match &context_path.0.get(i) {
+            Some(PathElement::Key(e, _)) => {
+                let mut found = false;
+                if e == ".." {
+                    while !found {
+                        j -= 1;
+                        if let Some(PathElement::Key(_, _)) = current_dir.0.get(j) {
+                            found = true;
+                        }
+                    }
+                    i += 1;
+                } else {
+                    break;
+                }
+            }
+            _ => break,
+        }
+    }
+
+    let mut return_path: Vec<PathElement> = current_dir.iter().take(j).cloned().collect();
+
+    context_path.iter().skip(i).for_each(|e| {
+        return_path.push(e.clone());
+    });
+    Path(return_path.into_iter().collect())
 }
```

Comment:
There is already a `path.join()`, I think this can be a `path.merge()`. But I'd also have it as `pub(crate) fn`.

Reply:
Even with my invented '..' syntax?

Reply:
You can give the `path.merge()` a `strategy` parameter to accommodate a the `..`. But what does the `..` actually mean? Why is context_path specifically different from other paths in the query planner?

Reply:
The .. is to indicate that it's a relative path rather than an absolute one. We're indicating that we need to traverse up the data tree from the current position.